pkg/cmd/psa/psa.go

```diff
@@ -0,0 +1,291 @@
+
+	// Print the violations.
+	return printViolations(podSecurityViolations)
+}
+
+// checkNamespacePodSecurity collects the pod security violations for a given
+// namespace on a stricter pod security enforcement.
+func (o *PSAOptions) checkNamespacePodSecurity(ns *corev1.Namespace) (*PodSecurityViolation, error) {
+	nsCopy := ns.DeepCopy()
+
+	// Get a higher enforcement value.
+	targetValue := ""
+	switch {
+	case o.level != "":
+		targetValue = o.level
+	case nsCopy.Labels[psapi.AuditLevelLabel] != "":
+		targetValue = nsCopy.Labels[psapi.AuditLevelLabel]
+	default:
+		targetValue = string(psapi.LevelRestricted)
+	}
+
+	// Update the pod security enforcement for the dry run.
+	nsCopy.Labels[psapi.EnforceLevelLabel] = targetValue
+
+	klog.V(4).Infof("Checking nsCopy %q for violations at level %q", nsCopy.Name, targetValue)
+
+	// Make a server-dry-run update on the nsCopy with the audit-level value.
+	_, err := o.client.CoreV1().
+		Namespaces().
+		Update(
+			context.Background(),
+			nsCopy,
+			metav1.UpdateOptions{DryRun: []string{"All"}},
+		)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get the warnings from the server-dry-run update.
+	warnings := o.warnings.PopAll()
+	if len(warnings) == 0 {
+		return nil, nil
+	}
+
+	return parseWarnings(warnings), nil
+}
+
+// getNamespaces returns the namespace that should be checked for pod security.
+// It could be given by the flag. Defaults to all namespaces.
+func (o *PSAOptions) getNamespaces() (*corev1.NamespaceList, error) {
+	if len(o.namespaces) == 0 {
+		namespaceList, err := o.client.CoreV1().
+			Namespaces().
+			List(context.Background(), metav1.ListOptions{})
+		if err != nil {
+			return nil, fmt.Errorf("failed to list namespaces: %w", err)
+		}
+
+		return namespaceList, nil
+	}
+
+	// Get the corev1.Namespace representation of the given namespaces.
+	// Also validate that those namespaces exist.
+	namespaceList := &corev1.NamespaceList{}
+	for _, namespace := range o.namespaces {
+		ns, err := o.client.CoreV1().
+			Namespaces().
+			Get(context.Background(), namespace, metav1.GetOptions{})
+		if err != nil {
+			return nil, fmt.Errorf("failed to get namespace %q: %w", namespace, err)
+		}
+
+		namespaceList.Items = append(namespaceList.Items, *ns)
+	}
+
+	return namespaceList, nil
+}
+
+// getPodController gets the deployment of a pod.
+func (o *PSAOptions) getPodController(pod *corev1.Pod) (any, error) {
+	parent := pod.ObjectMeta.OwnerReferences[0]
+
+	// If the pod is owned by a ReplicaSet, get the ReplicaSet's owner.
+	if parent.Kind == "ReplicaSet" {
+		replicaSet, err := o.client.AppsV1().
+			ReplicaSets(pod.Namespace).
+			Get(context.Background(), parent.Name, metav1.GetOptions{})
+		if err != nil {
+			return nil, fmt.Errorf("failed to get ReplicaSet %q: %w", parent.Name, err)
+		}
+
+		parent = replicaSet.ObjectMeta.OwnerReferences[0]
+	}
+
+	if _, ok := podControllers[parent.Kind]; !ok {
+		return nil, fmt.Errorf("pod controller %q is not supported", parent.Kind)
```

Comment:
no need to error.  just having the pod may help.

Reply:
I would then default to `nil, nil` in that case.